doc/reference/trd104-syscalls.md

```diff
@@ -303,61 +336,226 @@ call to invoke. The kernel invokes upcalls only in response to Yield
 
 | Argument               | Register |
 |------------------------|----------|
 | Yield number           | r0       |
-| No wait field          | r1       |
-| unused                 | r2       |
-| unused                 | r3       |
+| yield-result           | r1       |
+| yield-param-A          | r2       |
+| yield-param-B          | r3       |
 
+The Yield system call class has no return value. This is because
+invoking an upcall pushes that function call onto the stack, such
+that the return value of a call to yield system call may be the
+return value of the upcall.
 
-The yield number specifies which call is invoked.
+Yield variants which return a result to the caller will use one or more
+of the `yield-param-X` arguments to do so.  Allowing the kernel to pass
+a return value in a register back to userspace would require either
+re-entering the kernel or expensive execution architectures (e.g.,
+additional stacks or additional stack frames) for upcalls.
+
+The Yield number (in r0) specifies which call is invoked:
 
 | System call     | Yield number value |
 |-----------------|--------------------|
 | yield-no-wait   |                  0 |
 | yield-wait      |                  1 |
+| yield-wait-for  |                  2 |
 
+All other yield number values are reserved. If an invalid yield number
+is passed the kernel MUST return immediately and MUST NOT write to
+`yield-result`.
 
-All other yield number values are reserved. If an invalid
-yield number is passed the kernel MUST return immediately.
+`yield-result` contains the memory address of an 8-bit byte that Yield
+writes to indicate whether an upcall was invoked. If invoking Yield
+resulted in an upcall executing, Yield writes 1 to the field address. If
+invoking Yield resulted in no upcall executing, Yield writes 0 to the
+field address. Userspace SHOULD ensure that r1 points to a valid address
+in the current process. If userspace does not wish to recieve the Yield
+result, it SHOULD set r1 to `0x0`. The kernel SHALL write Yield result
+if r1 points to any valid process memory and SHALL NOT write Yield
+result if it points to an address not in the memory allocated to the
+calling process.
 
-The no wait field is only used by `yield-no-wait`. It contains the
-memory address of an 8-bit byte that `yield-no-wait` writes to
-indicate whether an upcall was invoked. If invoking `yield-no-wait`
-resulted in an upcall executing, `yield-no-wait` writes 1 to the
-field address. If invoking `yield-no-wait` resulted in no upcall
-executing, `yield-no-wait` writes 0 to the field address. This field
-allows userspace loops that want to flush the upcall queue to
-execute `yield-no-wait` until the queue is empty.
+**OPEN QUESTION:** ^^ This documents what we currently do, but it is the
+case right now that there is no way for userspace to identify an
+errorneous call to yield (e.g. invalid value in r0)---should instead
+`yeild-result` be written no matter what and an error case defined?
```

Comment:
I'm not sure what this means. What is an erroneous call to yield? Also, `yield-result` is written no matter what, assuming it is a valid process pointer. What would the error case capture? When are you going to enable spell check in your text editor :)?

Reply:
The only one I can think of is an illegal yield type, i.e. the invalid value in r0. FWIW, this is what the current TRD says:
<img width="1149" alt="image" src="https://github.com/tock/tock/assets/339422/35569853-f522-4a0c-84ec-69c6b3ff4977">
which basically kind of just doesn't say anything about what else does or doesn't happen if you make an impossible syscall.

Frankly, I don't really think it matters too much what we do, as so much else has gone wrong likely if you're asking for an invalid Yield, but I do feel like we should define what does or doesn't happen—especially if we have a "r1 is the yield-result" paradigm.

---

> When are you going to enable spell check in your text editor :)?

😬 , never... I learned a long time ago that spellcheck is too distracting for my brain to write/synthesize anything non-trivial. I do try to make a point of spellchecking non-draft PRs before pushing though :)

Reply:
> Frankly, I don't really think it matters too much what we do,

I think we are already clear on what happens:

> If userspace tries to invoke a system call that the kernel does not support, the system call will return a Failure result with an error code of NODEVICE or NOSUPPORT (Section 4).

Reply:
I think that a system call with class number 0 and r0==<anything not defined> is an unknown syscall and not an unknown yield.

Reply:
> I think that a system call with class number 0 and r0== is an unknown syscall and not an unknown yield.

The problem here is "how does the kernel tell userspace it was an invalid syscall?"

I think we the challenge with the current text / design is that these two parts are at odds:

> **the system call will return** a Failure result with an error code...

> The Yield system call class has no return value.

How can a system call class that has no return value return a Failure result?

We're in a pretty deep never-should-happen corner case here, but what I'm struggling with is 'what is the right thing for userspace to do'? Once userspace invokes a syscall with `svc 0`, i.e. something in the Yield class, it _cannot_ first 'read the return value to see if the syscall failed' as the Yield class defines different return semantics than every other syscall.